### DISCUSSION TITLE
MLX Roadmap?

### DISCUSSION THREAD

**Discussion by vade** [ml-explore/mlx#301] [General]
Hello

Since the announcement of MLX I've been excited and watching on the sidelines. I'm curious if there has been any public discussion of a roadmap for MLX?

I'm aware of some general efforts that seem quite useful

* Swift compatibility
* Quantization (recently merged)
* Expanding out supported layers and model types

All of this is awesome, but im left wondering a few things.

-  Who is MLX directly aimed at? IE: 
   - For researchers implementing new approaches within python wanting to take advantage of Apple Silicon? 
   - For developers looking take advantage of Apple silicon within applications or services using python or other languages?
   
- What language is the focus for MLX?
  - Given that MLX like most Pythonic libraries actually implements the most important code in C++, there is a C++ backend.
  - That backend however does not have a clearly documented API, nor are 'useful' abstractions easily composed with it.
  - Swift compatibility is being worked on, but to what end? Is Swift meant to be more like the C++ layer, where developers lose out on helpful higher level abstractions and composability, or is the goal of Swift to match some of Pythons higher level MLX Api?
  

I ask these questions from the perspective of an Application Developer building native apps for macOS and iOS. 

MLX seems very promising, but its unclear to me if I should invest my time learning MLX python and C++ API without understanding the overall goal of MLX writ large.

Thanks! I know this is early days and abstract, but I think having a public roadmap helps in a lot of ways. It can help guide contributions, help frame questions and answers and provide framing for how to accept PRs.

Thank you again

**Comment by dougdew64** [comment]
The MLX README offers this: 

`MLX is designed by machine learning researchers for machine learning researchers. The framework is intended to be user-friendly, but still efficient to train and deploy models. The design of the framework itself is also conceptually simple. We intend to make it easy for researchers to extend and improve MLX with the goal of quickly exploring new ideas.`

From that statement, I could imagine that having only the higher level python API would be sufficient to meet the needs of the target audience of MLX. In other words, it's not clear that either a Swift API or a higher level C++ API are necessary to serve the needs of machine learning researchers.

After thinking about all of this for a bit, it seems entirely reasonable that machine learning researchers would be the only users of MLX. The machine learning researchers would use MLX and the python API to train and deploy models. App developers would then consume those models by using Swift and the CoreML / MPS stack.

Rather than implement higher level C++ MLX APIs or Swift MLX APIs, perhaps a better use of time and effort is to ensure that MLX can export models for convenient consumption in the CoreML / MPS stack.

Rather than compete with the CoreML / MPS stack, MLX would complement the CoreML / MPS stack.

**Reply by RahulBhalley** [reply]
This might be the major goal. 

> The machine learning researchers would use MLX and the python API to train and deploy models. App developers would then consume those models by using Swift and the CoreML / MPS stack.

**Reply by dougdew64** [reply]
@RahulBhalley there is a much smaller constituency which probably won't be represented in the project goals: educators and authors.

My understanding is that you are an author. I'm not an author, but I am interested in creating supporting materials for authors.

For example, I'm currently interested in building atop MLX to create a SwiftUI-based LLM visualization such as this: https://bbycroft.net/llm. It drives me nuts that the explanations of LLMs which currently exist do not seem to bridge the gap between concepts and code. I believe that it is possible to do much better.

It seems to me that there might be some overlap between the needs of researchers and the needs of those who would explain LLMs to regular people such as me. For example, gathering statistics about sparsity of activations and perhaps even visualizing that sparsity. MLX + SwiftUI + MacOS seems like a combination which could serve as a basis for creating statistics-gathering and visualization apps which could be used by researchers and by authors. However, enabling that combination would require implementing a bunch of MLX Swift API code and probably also instrumenting the underlying MLX C++ code. And, it's not clear that that Swift API or that C++ instrumentation are aligned with this project's major goal(s).

**Reply by dougdew64** [reply]
How much better might some statistics, linear algebra, calculus or CS classes be if they were supplemented with motivating demonstrations such as visualizations of LLMs?

**Reply by dougdew64** [reply]
I wish that Xcode had custom debugger visualizers like Visual Studio does so that it would be possible to visualize LLM state while stepping through the MLX code. The custom debugger visualizers would provide opportunities to present helpful explanations along with the visualizations. Using a combination like that, it might be possible to fully bridge the gap between conceptual explanations and code.

**Reply by dougdew64** [reply]
Possibilities like that are the primary reason why I am so enthusiastic about this Apple-platform-specific MLX repo.

**Reply by vade** [reply]
https://developer.apple.com/library/archive/documentation/IDEs/Conceptual/CustomClassDisplay_in_QuickLook/CH01-quick_look_for_custom_objects/CH01-quick_look_for_custom_objects.html

**Reply by dougdew64** [reply]
I am ready for a new kind of book. One that integrates with a debugger.

**Reply by vade** [reply]
Can we keep this somewhat on topic please?

**Reply by dougdew64** [reply]
Thank you very much for sharing that url @vade !!!
